### PR TITLE
Add count parameter to schedules

### PR DESCRIPTION
**Pull request by wwajerowicz**

### CODE REVIEW

File: atlas-api/src/main/java/org/atlasapi/query/v4/schedule/ScheduleResolverBackedScheduleQueryExecutor.java [hunk 1]
```diff
@@ -59,7 +60,14 @@ public ScheduleResolverBackedScheduleQueryExecutor(ChannelResolver channelResolv
             throws QueryExecutionException {
         
         List<Channel> channels = resolveChannels(query);
-        ListenableFuture<Schedule> schedule = scheduleResolver.resolve(channels, query.getInterval(), query.getSource());
+        if(!query.getEnd().isPresent()) {
+            throw new UnsupportedOperationException("ScheduleResolverBackedScheduleQueryExecutor doesn't support 'count' queries. Please specify 'to' parameter in your request");
```
Comment (mbst-jamie) [inline atlas-api/src/main/java/org/atlasapi/query/v4/schedule/ScheduleResolverBackedScheduleQueryExecutor.java]: ScheduleResolverBackedScheduleQueryExecutor means nothing to 99% of people who will read that error message

File: atlas-api/src/main/java/org/atlasapi/query/v4/schedule/ScheduleRequestParser.java [hunk 2]
```diff
@@ -170,14 +188,32 @@ private Id transformToId(String channelId) throws QueryParseException {
         }
     }
 
-    private Interval extractInterval(HttpServletRequest request) {
+    private DateTime extractFrom(HttpServletRequest request) {
         DateTime now = clock.now();
         DateTime from = dateTimeParser.parse(getParameter(request, "from"), now);
```
Comment (tfm) [inline atlas-api/src/main/java/org/atlasapi/query/v4/schedule/ScheduleRequestParser.java]: extract param names to constants?

File: atlas-api/src/main/java/org/atlasapi/query/v4/schedule/ScheduleRequestParser.java [hunk 3]
```diff
@@ -170,14 +188,32 @@ private Id transformToId(String channelId) throws QueryParseException {
...
+    private void checkInterval(DateTime from, DateTime to) {
         Interval queryInterval = new Interval(from, to);
         checkArgument(!queryInterval.toDuration().isLongerThan(maxQueryDuration), "Query interval cannot be longer than %s", maxQueryDuration);
-        return queryInterval;
+    }
+
+    private Integer extractCount(HttpServletRequest request) {
+        Integer count = Integer.valueOf(getParameter(request, "count"));
+
+        checkArgument(COUNT_RANGE.contains(count),  "'count' must be in range %s, was %s", COUNT_RANGE, count);
```
Comment (tfm) [inline atlas-api/src/main/java/org/atlasapi/query/v4/schedule/ScheduleRequestParser.java]: "in range" yet you're not providing a range; perhaps "'count' must be at most %s"? Also, shouldn't that be %d?

File: atlas-api/src/main/java/org/atlasapi/query/v4/schedule/ScheduleResolverBackedScheduleQueryExecutor.java (hunk 1 above)
Comment (tfm) [inline atlas-api/src/main/java/org/atlasapi/query/v4/schedule/ScheduleResolverBackedScheduleQueryExecutor.java]: ^ that

File: atlas-api/src/test/java/org/atlasapi/query/v4/schedule/ScheduleRequestParserTest.java [hunk 4]
```diff
@@ -138,7 +143,56 @@ public void testCreatesSingleQueryFromValidSingleQueryStringWithNoExtension() th
...
         assertThat(query.getChannelId(), is(Id.valueOf(channel1.getId())));
-        assertThat(query.getInterval(), is(intvl));
+        assertThat(query.getStart(), is(start));
+        assertThat(query.getEnd().get(), is(end));
+        assertThat(query.getSource(), is(BBC));
+        assertThat(query.getContext().getAnnotations().forPath(ImmutableList.of(Resource.CONTENT)), is(Annotation.standard()));
+        assertThat(query.getContext().getApplicationSources(), is(sources));
+    }
+
+    @Test(expected=InvalidParameterException.class)
```
Comment (tfm) [inline atlas-api/src/test/java/org/atlasapi/query/v4/schedule/ScheduleRequestParserTest.java]: Can we also have a test for when count is too high? That would have caught the string format problem above.

**Comment by tfm** [comment]
back to you @wwajerowicz

File: atlas-api/src/main/java/org/atlasapi/query/v4/schedule/ScheduleRequestParser.java (hunk 3 above)
Comment (wwajerowicz) [inline atlas-api/src/main/java/org/atlasapi/query/v4/schedule/ScheduleRequestParser.java]: @tfm it is in a way a range, we don't want people to specify negative numbers or 0 as count.

Comment (wwajerowicz) [inline atlas-api/src/main/java/org/atlasapi/query/v4/schedule/ScheduleRequestParser.java]: And also `checkArgument` only accepts `%s` in the template. Which works fine with integers.

Comment (tfm) [inline atlas-api/src/main/java/org/atlasapi/query/v4/schedule/ScheduleRequestParser.java]: The error needs to say "between 0 and x", then

Comment (wwajerowicz) [inline atlas-api/src/main/java/org/atlasapi/query/v4/schedule/ScheduleRequestParser.java]: does it really make sense to accept 0 for the count?

Comment (tfm) [inline atlas-api/src/main/java/org/atlasapi/query/v4/schedule/ScheduleRequestParser.java]: good point, 1 :)

File: atlas-api/src/main/java/org/atlasapi/query/common/SetBasedRequestParameterValidator.java [hunk 5]
```diff
@@ -29,51 +30,35 @@ public Builder withOptionalParameters(String...parameters) {
             this.optional = ImmutableSet.copyOf(parameters);
             return this;
         }
+
+        public Builder withRequiredAlternativeParameters(String...parameters) {
```
Comment (tfm) [inline atlas-api/src/main/java/org/atlasapi/query/common/SetBasedRequestParameterValidator.java]: Space after ...

File: atlas-api/src/test/java/org/atlasapi/query/v4/schedule/ScheduleRequestParserTest.java (hunk 4 above)
Comment (wwajerowicz) [inline atlas-api/src/test/java/org/atlasapi/query/v4/schedule/ScheduleRequestParserTest.java]: Test added.

**Pull request merged**